packages/node/src/behaviors/thermostat/ThermostatServer.ts

```diff
@@ -4,14 +4,1000 @@
+                    nowUtc / 1000 + this.state.temperatureSetpointHoldDuration! * 60,
+                );
+            }
+        } else {
+            this.state.setpointHoldExpiryTimestamp = null;
+        }
+    }
+
+    #setupSetback() {
+        if (!this.features.setback) {
+            return;
+        }
+
+        this.reactTo(this.events.localTemperature$Changed, this.#handleTemperatureChangeForSetback);
+
+        this.reactTo(this.events.occupiedSetback$Changing, this.#assertOccupiedSetbackChanging);
+
+        if (this.features.occupancy) {
+            this.reactTo(this.events.unoccupiedSetback$Changing, this.#assertUnoccupiedSetbackChanging);
+        }
+    }
+
+    #assertOccupiedSetbackChanging(v: number | null) {
+        const croppedValue = this.#cropSetbackChange(v, "occupied");
+        if (v !== croppedValue) {
+            this.state.occupiedSetback = croppedValue;
+        }
+    }
+
+    #assertUnoccupiedSetbackChanging(v: number | null) {
+        const croppedValue = this.#cropSetbackChange(v, "unoccupied");
+        if (v !== croppedValue) {
+            this.state.unoccupiedSetback = croppedValue;
+        }
+    }
+
+    #handleTemperatureChangeForSetback(newValue: number | null) {
+        const setBackValue = this.occupied ? this.state.occupiedSetback : this.state.unoccupiedSetback;
+        if (newValue === null || setBackValue === null || setBackValue === 0) {
+            return;
+        }
+        const coolingSetPoint = this.coolingSetpoint;
+        const heatingSetPoint = this.heatingSetpoint;
+
+        // TODO ??
+        if (newValue > coolingSetPoint + setBackValue && this.coolingAllowed) {
+            this.adjustState(Thermostat.SystemMode.Cool);
+        } else if (newValue < heatingSetPoint - setBackValue && this.heatingAllowed) {
+            this.adjustState(Thermostat.SystemMode.Heat);
+        }
+    }
+
+    protected get heatingAllowed() {
+        return (
+            this.features.heating &&
+            ![
+                Thermostat.ControlSequenceOfOperation.CoolingOnly,
+                Thermostat.ControlSequenceOfOperation.CoolingAndHeatingWithReheat,
+            ].includes(this.internal.controlSequenceOfOperation)
+        );
+    }
+
+    protected get coolingAllowed() {
+        return (
+            this.features.cooling &&
+            ![
+                Thermostat.ControlSequenceOfOperation.HeatingOnly,
+                Thermostat.ControlSequenceOfOperation.HeatingWithReheat,
+            ].includes(this.internal.controlSequenceOfOperation)
+        );
+    }
+
+    /**
+     * Adjust the state of the thermostat based on the new system mode.
+     * When the AutoMode feature is enabled and the current system mode is Auto, the thermostat running mode
+     * is adjusted instead of the system mode.
+     */
+    protected adjustState(newState: Thermostat.SystemMode) {
+        if (this.features.autoMode && this.state.systemMode === Thermostat.SystemMode.Auto) {
```

Comment:
Lots of redundancy here.  Seems like the if statement should just select the state attribute and enum & place in variables.  Then everything else is the and